Source/Controllers/Other/SecureBookmarkManager.swift

```diff
@@ -0,0 +1,308 @@
+    }
+
+    /// revokeBookmark
+    ///  - Parameters:
+    ///     - filename: filename to revoke secure bookmark for
+    /// - Returns: Bool on success or fail
+    @objc func revokeBookmark(filename: String) -> Bool {
+        for (index, bookmarkDict) in bookmarks.enumerated() {
+            for (key, urlData) in bookmarkDict {
+                if key == filename {
+                    do {
+                        os_log("Revoking bookmark for: %@", log: log, type: .debug, filename)
+                        Crashlytics.crashlytics().log("Revoking bookmark for: \(filename)")
+
+                        os_log("bookmarks[%i]: %@", log: log, type: .debug, index, bookmarks[index])
+
+                        var bookmarkDataIsStale = false
+
+                        // need to get the proper URL
+                        let spData = SecureBookmark.getDecodedData(encodedData: urlData)
+
+                        os_log("Attempting to resolve bookmark data for %@", log: log, type: .debug, key)
+                        Crashlytics.crashlytics().log("Attempting to resolve bookmark data for: \(key)")
+                        // always resolve with just URLBookmarkResolutionWithSecurityScope
+                        let urlForBookmark = try URL(resolvingBookmarkData: spData.bookmarkData, options: [URLBookmarkResolutionWithSecurityScope], relativeTo: nil, bookmarkDataIsStale: &bookmarkDataIsStale)
+
+                        // you do not need to call .stopAccessingSecurityScopedResource()
+                        // simply remove from bookmarks
+
+                        resolvedBookmarks.removeAll(where: { $0 == urlForBookmark })
+                        bookmarks.remove(at: index)
+                        iChangedTheBookmarks = true
+                        prefs.set(bookmarks, forKey: SASecureBookmarks)
+                        os_log("Successfully revoked bookmark for: %@", log: log, type: .debug, filename)
+                        Crashlytics.crashlytics().log("Successfully revoked bookmark for: \(filename)")
+                        return true
+                    } catch {
+                        os_log("Error resolving bookmark: key = %@. Error: %@", log: log, type: .error, key, error.localizedDescription)
+                        Crashlytics.crashlytics().log("Error resolving bookmark: key = \(key). Error: \(error.localizedDescription)")
+                        os_log("Failed to revoke bookmark for: %@", log: log, type: .debug, filename)
+                        Crashlytics.crashlytics().log("Failed to revoke bookmark for: \(filename)")
+                        return false
+                    }
+                }
+            }
+        }
+
+        os_log("Failed to revoke bookmark for: %@", log: log, type: .debug, filename)
+        Crashlytics.crashlytics().log("Failed to revoke bookmark for: \(filename)")
+        return false
+    }
+
+    // revoke secure access to all bookmarks
+    @objc func stopAllSecurityScopedAccess() {
```

Comment:
This is what concerns me here - stopping access for everything using the same instance.
Perhaps instead the instance could somehow keep track of what accessed what and only stop accessing things requested by a specific class. In other words, instead of one global resolved bookmarks buffer, some sort of reference counting or per-class buffer.

Alternatively, if it's okay if we keep bookmarks accessed while the app is running, we could just keep resolving un-resolved bookmarks as we go and stop accessing all of them when the app is closed/when the last window is closed? Would that be acceptable to macOS? That way we'd be sure nothing was accessing a bookmark when we de-resolved it

Reply:
as above - will review the code and see where we request and revoke access. macOS only gets unhappy if you have 1000s of bookmarks .. like near `sysctl kern.maxfiles`

Reply:
ok ... we request access to all bookmarks on startup, then add more if user chooses a file location, or revoke from the Files pref screen. But we only revoke at file basis, and actually don't even call stopAccess .. I was seeing errors.

stopAllSecurityScopedAccess is called from the file/network pref pane dealloc methods ... I haven't seen them called yet.

Other locations are the connectionController dealloc and `_documentWillClose`. I can't trigger `_documentWillClose`. It runs when a DatabaseDocument object is dealloc'd. Again, can't trigger it.